### PR TITLE
Add support for reference to public static properties in settings

### LINKED ISSUE #1057: KeyValuePairSettings : support for static members via `::`
Related to discussion going on in serilog/serilog-sinks-console#16

It would be nice to support **static members** for configuration methods that accept a complex type .

The typical use case is *Serilog.Sinks.Console* with its configuration method : 
```csharp
public static LoggerConfiguration Console(
	this LoggerSinkConfiguration sinkConfiguration,
	LogEventLevel restrictedToMinimumLevel = LevelAlias.Minimum,
	string outputTemplate = DefaultConsoleOutputTemplate,
	IFormatProvider formatProvider = null,
	LoggingLevelSwitch levelSwitch = null,
	LogEventLevel? standardErrorFromLevel = null,
	ConsoleTheme theme = null)
```
where `ConsoleTheme` is an `abstract class`, but static members exist (`ConsoleTheme.None`, `SystemConsoleTheme.Literate`, `AnsiConsoleTheme.Grayscale` [etc](https://github.com/serilog/serilog-sinks-console#themes) )

One possible syntax that would be do access static members via `::` (like in [Powershell](https://docs.microsoft.com/en-us/powershell/scripting/getting-started/cookbooks/using-static-classes-and-methods?view=powershell-5.1#displaying-static-properties-of-systemenvironment) or [PHP](http://php.net/manual/en/language.oop5.static.php#language.oop5.static.properties)), for instance 
```xml
<configuration>
  <appSettings>
    <add key="serilog:using:Console" value="Serilog.Sinks.Console" />
    <add key="serilog:write-to:Console.theme" value="Serilog.Sinks.SystemConsole.Themes.AnsiConsoleTheme::Grayscale, Serilog.Sinks.Console" />
```

would use `static` property `Grayscale` of type `Serilog.Sinks.SystemConsole.Themes.AnsiConsoleTheme` in assembly `Serilog.Sinks.Console`.

As in #1056 the idea is that we should be able to specify the minimum information about the type to make the configuration less verbose (for instance, no need for the assembly name, Version, PublicKeyToken etc if we can avoid it).

Other use-case where this would help : *Serilog.Sinks.Email* that accepts a `EmailConnectionInfo` for advanced configuration

### PR DESCRIPTION
**Pull request by tsimbalar**
as a value for parameters, using the syntax `NameSpace.To.ConcreteType::StaticProperty, AssemblyName`

**What issue does this PR address?**
resolves #1057 

**Does this PR introduce a breaking change?**
It shouldn't, although technically, there is a chance that "string" settings whose value looks like `"Namespace.To.TypeName::PropertyName"` may now be interpreted as a reference to a static property, even though I've done my best to avoid ambiguities.  

**Please check if the PR fulfills these requirements**
- [x] The commit follows our [guidelines](https://github.com/serilog/serilog/blob/dev/CONTRIBUTING.md)
- [x] Unit Tests for the changes have been added (for bug fixes / features)

**Other information**:
When / if accepted, similar code should be needed in *Serilog.Settings.Configuration* to handle issue serilog/serilog-settings-configuration#73 .

### CODE REVIEW

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs [hunk 1]
```diff
@@ -29,6 +35,22 @@ class SettingValueConversions
 
         public static object ConvertToType(string value, Type toType)
         {
+            if (TryParseStaticMemberAccessor(value, out var accessorTypeName, out var memberName))
```
Comment (nblumhardt) [inline src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: Perhaps this could move down to the `if ((toTypeInfo.IsInterface || toTypeInfo.IsAbstract) && !string.IsNullOrWhiteSpace(value))`  block? If it's first up, here, we might hit issues where this is ambiguous, e.g. the target `Type` is `string`.

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs [hunk 2]
```diff
@@ -29,6 +35,22 @@ class SettingValueConversions
...
+            {
+                var accessorType = Type.GetType(accessorTypeName, throwOnError: true);
```
Comment (skomis-mm) [inline src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: Probably, for a better UX could we use full type names without specifying assembly? We could look for the target type in configuration assemblies (`using` section) something like:
```csharp
var accessorType = Type.GetType(typeName) ??
            _assemblies.SelectMany(a => a.ExportedTypes)
                        .FirstOrDefault(t => t.FullName == typeName);
```

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs [hunk 3]
```diff
@@ -29,6 +35,22 @@ class SettingValueConversions
...
+                var publicStaticPropertyInfo = accessorType.GetTypeInfo().DeclaredProperties
+                    .Where(x => x.Name == memberName)
```
Comment (skomis-mm) [inline src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: Perhaps we could also add support for the public static fields at the same time?

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs (hunk 2 above)
Comment (nblumhardt) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: If we do that for members, could/should we do the same thing for instances?

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs (hunk 1 above)
Comment (tsimbalar) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: That was my initial implementation, but then I thought it would be strange to limit it to only `abstract class`es and `interface`s, because after all this can also be useful in other cases, like for sinks that accept *complex* parameters that are not `abstract` nor `interface` (for instance : `EmailConnectionInfo` in *Serilog.Sinks.Email* , `ColumnOptions` for *Serilog.Sinks.MSSqlServer* ...) 

This also gives a workaround for configuration that canot be easily defined with the keay-value settings syntax, by allowing to define it as code in a static property.

Also, I've tried to make the *regex* for static-property accessor as specific as possible to avoid false positives.  

What do you think ?

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs (hunk 3 above)
Comment (tsimbalar) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: I have mixed feelings about accessing arbitrary properties, but I guess if they are `public`, they are part of the public API so technically accessible. I'll include public properties as well

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs (hunk 2 above)
Comment (tsimbalar) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: For now I've tried to not deviate too much from the way arbitrary sub-types (for interfaces and abstract classes) are referenced, that is with full assembly-qualified type names. Maybe lifting that restriction would be nice, but I would keep it as a separate issue / PR.

Comment (nblumhardt) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: Sounds good

File: src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs (hunk 1 above)
Comment (tsimbalar) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: @nblumhardt any opinion on the subject? :)

Comment (nblumhardt) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: I think it's a fair consideration - just not sure if it's behavior we want to commit to yet in this PR. (We'd also then need to consider whether assembly-qualified type names should be supported in more places, since subclassing is a possible workaround for other scenarios, plus look at how this would interact with some kind of "structure" syntax for populating fields of complex configuration parameters down the track).

TL;DR: I don't think it's a bad idea - just might be scope creep for this PR, since we haven't really dug through all the possible implications :-) .. what do you think?

Comment (tsimbalar) [reply src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs]: That makes sense 👍 
Easier to make it support a few use cases initially and extend it later on if needed.

Thanks !

**Comment by tsimbalar** [comment]
I've updated my PR with the suggested changes.

**Comment by skomis-mm** [comment]
👍
